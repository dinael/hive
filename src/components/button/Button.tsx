import { FC, MouseEventHandler, Children, ReactNode } from 'react';
import styles from './Button.module.scss'
import Icon from '@/components/icon/Icon'

export type ButtonProps = {
  type?: 'button' | 'submit' | 'reset' | 'link'
  variant?: 'primary' | 'secondary' | 'alt' | 'ghost' | 'success' | 'danger'
  kind?: 'default' | 'rounded' | 'fullwidth'
  size?: 'default' | 'xs' | 's' | 'm' | 'l'
  text?: string
  ellipsis?: boolean
  icon?: string
  iconOnly?: boolean
  iconPosition?: 'left' | 'right'
  url?: string
  description?: string
  disabled?: boolean
  id?: string
  className?: string
  children?: ReactNode;
  onClick?: MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>
}

export const Button: FC<ButtonProps> = ({
  type = 'button',
  variant = 'primary',
  kind = 'default',
  size,
  text = 'button',
  ellipsis = false,
  icon,
  iconOnly = false,
  iconPosition = 'left',
  url,
  description: a11yText = 'a11y Button',
  disabled = false,
  className,
  id,
  children,
  ...props
}: ButtonProps) => {

  function handleClick(e:any) {
    e.preventDefault()
  }

  const buttonStyles = `
    ${styles['button-component']}
    ${styles[variant ? variant : 'primary']}
    ${styles[size ? size : 'default']}
    ${styles[kind ? kind : 'default']}
    ${styles[iconOnly || kind === 'rounded' ? 'icon-only' : iconPosition]}
    ${disabled ? styles['disabled'] : ''}
    ${className || ''}
  `

  const iconStyles = `${styles['icon-button']}`

  const textStyles = `
    ${styles.text}
    ${ellipsis ? styles['ellipsis'] : ''}
    ${iconOnly || kind === 'rounded' ? 'sr-only' : ''}
  `

  const buttonIcon = icon ? <Icon name={icon} size={size} className={iconStyles} /> : null

  const buttonText = !iconOnly && kind !== 'rounded' ? (<span className={textStyles}>{text}</span>) : null

  const buttonContent = (
    <>
      {buttonIcon}
      {buttonText}
    </>
  );

  if (type !== 'link') {
    return (
      <button
        id={id}
        type={type}
        aria-label={a11yText}
        disabled={disabled}
        className={buttonStyles}
        onClick={handleClick}
        {...props}>
        {children ? children : buttonContent}
      </button>
    )
  }

  return (
    <a
      id={id}
      href={url}
      target="_self"
      rel="noopener noreferrer"
      aria-label={a11yText}
      className={buttonStyles}
      onClick={handleClick}
      {...props}>
      {children ? children : buttonContent}
    </a>
  )
}

export default Button

