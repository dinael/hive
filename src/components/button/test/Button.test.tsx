import { describe, test, expect, vi} from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import { Button, ButtonProps } from '../Button'

describe('Button', () => {
  const defaultProps: ButtonProps = {
    type: 'button',
    text: 'Button text',
    variant: 'primary',
    kind: 'default',
    size: 'xs',
    ellipsis: false,
    icon: 'close',
    iconOnly: false,
    iconPosition: 'left',
    url: 'https://www.example.com',
    description: 'Text for a11y',
    disabled: false,
    className: 'className-test',
    onClick: vi.fn(),
  }

  test('Button renders snapshot', () => {
    const { container } = render(<Button {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('Button renders correctly', () => {
    const { container } = render(<Button {...defaultProps} />)
    const buttonElement = container.querySelector('button')
    expect(buttonElement?.tagName).toBe('BUTTON')
    expect(container).toMatchSnapshot()
  })

  test('Button renders text correctly', () => {
    const { container } = render(<Button {...defaultProps} />)
    const buttonElement = container.querySelector('button span:nth-child(2)')
    expect(buttonElement?.textContent).toBe(defaultProps.text)
    expect(screen.getByText(`${defaultProps.text}`))
    expect(container).toMatchSnapshot()
  })

  test('Button renders icon correctly', () => {
    const { container } = render(<Button {...defaultProps} />)
    const buttonElement = container.querySelector('button span')
    expect(buttonElement?.className).toContain('icon-component')
    expect(container).toMatchSnapshot()
  })

  test('Button renders as link correctly', () => {
    const { container } = render(<Button {...defaultProps} type={'link'} url={defaultProps.url} />)
    const buttonElement = container.querySelector('a')
    expect(buttonElement?.tagName).toBe('A')
    expect(buttonElement?.getAttribute('href')).toBe(defaultProps.url)
    expect(container).toMatchSnapshot()
  })

  test('Button onClick', () => {
    const onClick = vi.fn()
    const { container } = render(<Button {...defaultProps} onClick={onClick} />)
    const buttonElement = container.querySelector('button');
    if (buttonElement) {
      fireEvent.click(buttonElement);
    }
    expect(onClick).toHaveBeenCalled();
    expect(container).toMatchSnapshot()
  })
})
