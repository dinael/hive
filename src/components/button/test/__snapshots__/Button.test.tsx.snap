// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > Button onClick 1`] = `
<div>
  <button
    aria-label="Text for a11y"
    class="
    button-component
    primary
    xs
    default
    left
    
    className-test
  "
    type="button"
  >
    <span
      aria-hidden="true"
      class="
    icon-component
    xs
    icon-button
  "
    >
      <svg
        viewBox="0 0 24 24"
      >
        <title>
          close
        </title>
        <use
          xlink:href="/src/assets/sprite.svg#close"
        />
      </svg>
    </span>
    <span
      class="
    text
    
    
  "
    >
      Button text
    </span>
  </button>
</div>
`;

exports[`Button > Button renders as link correctly 1`] = `
<div>
  <a
    aria-label="Text for a11y"
    class="
    button-component
    primary
    xs
    default
    left
    
    className-test
  "
    href="https://www.example.com"
    rel="noopener noreferrer"
    target="_self"
  >
    <span
      aria-hidden="true"
      class="
    icon-component
    xs
    icon-button
  "
    >
      <svg
        viewBox="0 0 24 24"
      >
        <title>
          close
        </title>
        <use
          xlink:href="/src/assets/sprite.svg#close"
        />
      </svg>
    </span>
    <span
      class="
    text
    
    
  "
    >
      Button text
    </span>
  </a>
</div>
`;

exports[`Button > Button renders children element correctly 1`] = `
<div>
  <button
    aria-label="a11y Button"
    class="
    button-component
    primary
    default
    default
    left
    
    
  "
    type="button"
  >
    text
  </button>
</div>
`;

exports[`Button > Button renders correctly 1`] = `
<div>
  <button
    aria-label="Text for a11y"
    class="
    button-component
    primary
    xs
    default
    left
    
    className-test
  "
    type="button"
  >
    <span
      aria-hidden="true"
      class="
    icon-component
    xs
    icon-button
  "
    >
      <svg
        viewBox="0 0 24 24"
      >
        <title>
          close
        </title>
        <use
          xlink:href="/src/assets/sprite.svg#close"
        />
      </svg>
    </span>
    <span
      class="
    text
    
    
  "
    >
      Button text
    </span>
  </button>
</div>
`;

exports[`Button > Button renders icon correctly 1`] = `
<div>
  <button
    aria-label="Text for a11y"
    class="
    button-component
    primary
    xs
    default
    left
    
    className-test
  "
    type="button"
  >
    <span
      aria-hidden="true"
      class="
    icon-component
    xs
    icon-button
  "
    >
      <svg
        viewBox="0 0 24 24"
      >
        <title>
          close
        </title>
        <use
          xlink:href="/src/assets/sprite.svg#close"
        />
      </svg>
    </span>
    <span
      class="
    text
    
    
  "
    >
      Button text
    </span>
  </button>
</div>
`;

exports[`Button > Button renders snapshot 1`] = `
<div>
  <button
    aria-label="Text for a11y"
    class="
    button-component
    primary
    xs
    default
    left
    
    className-test
  "
    type="button"
  >
    <span
      aria-hidden="true"
      class="
    icon-component
    xs
    icon-button
  "
    >
      <svg
        viewBox="0 0 24 24"
      >
        <title>
          close
        </title>
        <use
          xlink:href="/src/assets/sprite.svg#close"
        />
      </svg>
    </span>
    <span
      class="
    text
    
    
  "
    >
      Button text
    </span>
  </button>
</div>
`;

exports[`Button > Button renders text correctly 1`] = `
<div>
  <button
    aria-label="Text for a11y"
    class="
    button-component
    primary
    xs
    default
    left
    
    className-test
  "
    type="button"
  >
    <span
      aria-hidden="true"
      class="
    icon-component
    xs
    icon-button
  "
    >
      <svg
        viewBox="0 0 24 24"
      >
        <title>
          close
        </title>
        <use
          xlink:href="/src/assets/sprite.svg#close"
        />
      </svg>
    </span>
    <span
      class="
    text
    
    
  "
    >
      Button text
    </span>
  </button>
</div>
`;
